source/opt/dead_branch_elim_pass.cpp

```diff
@@ -185,20 +186,57 @@ bool DeadBranchElimPass::GetSelectionBranch(ir::BasicBlock* bp,
   return true;
 }
 
-bool DeadBranchElimPass::HasNonPhiRef(uint32_t labelId) {
+bool DeadBranchElimPass::HasNonPhiNonBackedgeRef(uint32_t labelId) {
   analysis::UseList* uses = def_use_mgr_->GetUses(labelId);
   if (uses == nullptr)
     return false;
-  for (auto u : *uses)
-    if (u.inst->opcode() != SpvOpPhi)
+  for (auto u : *uses) {
+    if (u.inst->opcode() != SpvOpPhi &&
+        backedges_.find(u.inst) == backedges_.end())
       return true;
+  }
   return false;
 }
 
+void DeadBranchElimPass::ComputeBackEdges(
+    std::list<ir::BasicBlock*>& structuredOrder) {
+  backedges_.clear();
+  std::unordered_set<uint32_t> visited;
+  // In structured order, edges to visited blocks are back edges
+  for (auto bi = structuredOrder.begin(); bi != structuredOrder.end(); ++bi) {
+    auto ii = (*bi)->end();
+    --ii;
+    switch(ii->opcode()) {
+      case SpvOpBranch: {
+        const uint32_t labId = ii->GetSingleWordInOperand(
+            kBranchTargetLabIdInIdx);
+        if (visited.find(labId) != visited.end())
+          backedges_.insert(&*ii);
+      } break;
+      case SpvOpBranchConditional: {
+        const uint32_t tLabId = ii->GetSingleWordInOperand(
+            kBranchCondTrueLabIdInIdx);
+        if (visited.find(tLabId) != visited.end()) {
+          backedges_.insert(&*ii);
+          break;
+        }
+        const uint32_t fLabId = ii->GetSingleWordInOperand(
+            kBranchCondFalseLabIdInIdx);
+        if (visited.find(fLabId) != visited.end())
+          backedges_.insert(&*ii);
+      } break;
+      default:
+        break;
+    }
+    visited.insert((*bi)->id());
```

Comment:
I think you need to insert into visited at the start of the loop.  Otherwise you will miss back edges of loops that contain just a single basic block.

Reply:
Agreed.